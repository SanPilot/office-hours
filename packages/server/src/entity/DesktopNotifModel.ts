import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  ManyToOne,
  JoinColumn,
  AfterInsert,
  AfterLoad,
} from "typeorm";
import { UserModel } from "./UserModel";
import { notifyUser } from "../api/notifRoutes";

@Entity("desktop_notif_model")
export class DesktopNotifModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  endpoint: string;

  @Column({ nullable: true })
  expirationTime: Date;

  @Column("text")
  p256dh: string;

  @Column("text")
  auth: string;

  @ManyToOne((type) => UserModel, (user) => user.desktopNotifs)
  @JoinColumn({ name: "userId" })
  user: UserModel;

  @Column({ nullable: true })
  userId: number;

  @AfterInsert()
  async initialNotifyUser() {
    await notifyUser(
      this,
      "You've successfully signed up for desktop notifications!"
    );
  }

  // TODO: add functionality to notify user once they hit top of queue (probably have to make a
  // afterload listener on Queue,a nd then try to find the notifmodel of the user there, and then
  // just do something like await notifyUser(queue.user.notifs, "joe mama"))
}
