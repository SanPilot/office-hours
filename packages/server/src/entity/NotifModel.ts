import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  ManyToOne,
  JoinColumn,
  OneToMany,
} from "typeorm";
import { UserCourseModel } from "./UserCourseModel";
import { NotifBody } from "@template/common";
import { UserModel } from "./UserModel";

/*
based off of this from the web push front end 
{
  endpoint: 'https://fcm.googleapis.com/fcm/send/d_OpZBBSK4g:APA91bFPDE55-Q_-emManQ1HaoU8uMFRLCyfxIldgRHpQMr7X1bG2z6PW5RrTpu5FCCvbjOgw_VXdDmkp1h5OLYkKiWK1EM3pWiq1kGijD3ieer7cuP5Pap6NT_zqMWB1d1XehYOBrLS',
  expirationTime: null,
  keys: {
    p256dh: 'BBmo6-uI-no3F_PRuGe-zV-kw-jZh4jWH_XI0SGb2LnqcVDTY2nuaF50_5nWYufcwGqXyxy_bfVqrJ1n_3-fgfo',
  }
}
*/

@Entity("notif_model")
export class NotifModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  endpoint: string;

  @Column({ nullable: true })
  expirationTime: Date;

  @Column("text")
  p256dh: string;

  @Column("text")
  auth: string;

  @ManyToOne((type) => UserModel, (user) => user.notifs)
  @JoinColumn({ name: "userId" })
  user: Promise<UserModel>;

  @Column({ nullable: true })
  userId: number;
}
