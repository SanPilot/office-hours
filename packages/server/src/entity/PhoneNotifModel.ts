import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  ManyToOne,
  JoinColumn,
} from "typeorm";
import { UserModel } from "./UserModel";

@Entity("phone_notif_model")
export class PhoneNotifModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  phoneNumber: string;

  @ManyToOne((type) => UserModel, (user) => user.phoneNotifs)
  @JoinColumn({ name: "userId" })
  user: UserModel;

  @Column({ nullable: true })
  userId: number;
}
