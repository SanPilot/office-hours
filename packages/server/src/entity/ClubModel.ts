import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  OneToMany,
} from "typeorm";
import { OfficeHourModel } from "./OfficeHourModel";

@Entity("club")
export class ClubModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  name: string;

  @Column()
  rating: number;
}
