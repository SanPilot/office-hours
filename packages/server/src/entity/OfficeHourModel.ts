import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  ManyToOne,
  JoinColumn,
} from "typeorm";
import { CourseModel } from "./CourseModel";

@Entity("office_hour")
export class OfficeHourModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => CourseModel, (course) => course.officeHours)
  @JoinColumn({ name: "courseId" })
  course: Promise<CourseModel>;

  @Column("text")
  title: string;

  @Column("text")
  room: string;

  @Column()
  startTime: Date;

  @Column()
  endTime: Date;
}
