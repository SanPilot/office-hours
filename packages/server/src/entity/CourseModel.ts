import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  OneToMany,
} from "typeorm";
import { OfficeHourModel } from "./OfficeHourModel";
import { QueueModel } from "./QueueModel";

/**
 * Represents a course in the context of office hours.
 * @param id - The id number of this Course.
 * @param name - The subject and course number of this course. Ex: "CS 2500"
 * @param semester - The semester of this course.
 */
/*interface Course {
    id: number;
    name: string;
    url: string;
    semester: Semester;
    users: UserCourse[]
}*/

@Entity("course_model")
export class CourseModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @OneToMany((type) => OfficeHourModel, (oh) => oh.course)
  officeHours: Promise<OfficeHourModel[]>;

  @OneToMany((type) => QueueModel, (q) => q.course)
  whiteBoard: Promise<QueueModel[]>;

  @Column("text")
  name: string;

  @Column("text")
  icalUrl: string;

  //todo: add semester + userCourse
}
