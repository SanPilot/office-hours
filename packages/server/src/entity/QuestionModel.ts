import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  BaseEntity,
  ManyToOne,
  JoinColumn,
} from "typeorm";
import { UserCourseModel } from "./UserCourseModel";
import { QueueModel } from "./QueueModel";
import { QuestionType, QuestionStatus } from "@template/common";

@Entity("question_model")
export class QuestionModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => QueueModel, (q) => q.questions)
  @JoinColumn({ name: "queueId" })
  queue: Promise<QueueModel>;

  @Column({ nullable: true })
  queueId: number;

  @Column("text")
  text: string;

  @ManyToOne((type) => UserCourseModel)
  @JoinColumn({ name: "creatorId" })
  creator: Promise<UserCourseModel>;

  @Column({ nullable: true })
  creatorId: number;

  @ManyToOne((type) => UserCourseModel)
  @JoinColumn({ name: "taHelpedId" })
  taHelped: Promise<UserCourseModel>;

  @Column({ nullable: true })
  taHelpedId: number;

  @Column()
  createdAt: Date;

  @Column({ nullable: true })
  helpedAt: Date;

  @Column({ nullable: true })
  closedAt: Date;

  @Column("text")
  questionType: QuestionType;

  @Column("text")
  status: QuestionStatus;
}
