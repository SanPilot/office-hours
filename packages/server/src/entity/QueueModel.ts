import {
  BaseEntity,
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from "typeorm";
import { CourseModel } from "./CourseModel";
import { QuestionModel } from "./QuestionModel";

@Entity("queue_model")
export class QueueModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => CourseModel, (course) => course.queues)
  @JoinColumn({ name: "courseId" })
  course: CourseModel;

  @Column({ nullable: true })
  courseId: number;

  @Column("text")
  room: string;

  @OneToMany((type) => QuestionModel, (qm) => qm.queue)
  questions: QuestionModel[];

  @Column("text", { nullable: true })
  notes: string;

  // TODO: eventually figure out how staff get sent to FE as well
}
