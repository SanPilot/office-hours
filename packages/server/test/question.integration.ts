import {
  ClosedQuestionStatus,
  OpenQuestionStatus,
  QuestionStatusKeys,
  QuestionType,
} from '@koh/common';
import { UserModel } from 'profile/user.entity';
import { QueueModel } from 'queue/queue.entity';
import supertest from 'supertest';
import { QuestionModel } from '../src/question/question.entity';
import { QuestionModule } from '../src/question/question.module';
import {
  CourseFactory,
  QuestionFactory,
  QueueFactory,
  StudentCourseFactory,
  TACourseFactory,
  UserCourseFactory,
  UserFactory,
  OfficeHourFactory,
  ClosedOfficeHourFactory,
} from './util/factories';
import {
  expectUserNotified,
  modifyMockNotifs,
  setupIntegrationTest,
} from './util/testUtils';

describe('Question Integration', () => {
  const supertest = setupIntegrationTest(QuestionModule, modifyMockNotifs);

  describe('GET /questions/:id', () => {
    it('gets a question with the given id', async () => {
      const course = await CourseFactory.create();
      const q = await QuestionFactory.create({
        text: 'Help pls',
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
        queue: await QueueFactory.create({
          course: course,
        }),
      });

      await UserCourseFactory.create({
        user: await UserFactory.create(),
        course: course,
      });

      const response = await supertest({ userId: 99 })
        .get(`/questions/${q.id}`)
        .expect(200);
      expect(response.body).toMatchSnapshot();
    });
    it('fails to get a non-existent question', async () => {
      await supertest({ userId: 99 })
        .get(`/questions/999`)
        .expect(404);
    });
  });

  describe('POST /questions', () => {
    const postQuestion = (
      user: UserModel,
      queue: QueueModel,
      force = false,
    ): supertest.Test =>
      supertest({ userId: user.id })
        .post('/questions')
        .send({
          text: "Don't know recursion",
          questionType: QuestionType.Concept,
          queueId: queue.id,
          force,
        });

    it('posts a new question', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({
        course: course,
        allowQuestions: true,
      });
      const user = await UserFactory.create();
      await StudentCourseFactory.create({ user, courseId: queue.courseId });
      expect(await QuestionModel.count({ where: { queueId: 1 } })).toEqual(0);
      const response = await postQuestion(user, queue).expect(201);
      expect(response.body).toMatchObject({
        text: "Don't know recursion",
        helpedAt: null,
        closedAt: null,
        questionType: 'Concept',
        status: 'Drafting',
      });
      expect(await QuestionModel.count({ where: { queueId: 1 } })).toEqual(1);
    });
    it('ta cannot post  a new question', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({
        course: course,
        allowQuestions: true,
      });
      const user = await UserFactory.create();
      await TACourseFactory.create({ user, courseId: queue.courseId });
      expect(await QuestionModel.count({ where: { queueId: 1 } })).toEqual(0);
      await postQuestion(user, queue).expect(401);
    });
    it('post question fails with non-existent queue', async () => {
      await supertest({ userId: 99 })
        .post('/questions')
        .send({
          text: "Don't know recursion",
          questionType: QuestionType.Concept,
          queueId: 999,
          force: false,
        })
        .expect(404);
    });
    it('does not allow posting in course student is not in', async () => {
      const queueImNotIn = await QueueFactory.create({
        allowQuestions: true,
      });
      const user = await UserFactory.create();
      await StudentCourseFactory.create({
        user,
        course: await CourseFactory.create(),
      });
      await postQuestion(user, queueImNotIn).expect(404);
    });
    it('post question fails on closed queue', async () => {
      const officeHours = await ClosedOfficeHourFactory.create();
      const course = await CourseFactory.create({
        officeHours: [officeHours],
      });

      const queue = await QueueFactory.create({
        course: course,
        officeHours: [officeHours],
        allowQuestions: true,
      });
      expect(await queue.checkIsOpen()).toBe(false);

      const user = await UserFactory.create();
      await StudentCourseFactory.create({ user, courseId: queue.courseId });
      await supertest({ userId: user.id });
      await postQuestion(user, queue).expect(400);
    });
    it('post question fails with bad params', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({
        courseId: course.id,
        allowQuestions: true,
      });
      const user = await UserFactory.create();
      await StudentCourseFactory.create({ user, courseId: queue.courseId });
      await supertest({ userId: user.id })
        .post('/questions')
        .send({
          text: 'I need help',
          questionType: 'bad param!',
          queueId: 1, // even with bad params we still need a queue
          force: false
        })
        .expect(400);
    });
    it("can't create more than one open question at a time", async () => {
      const course = await CourseFactory.create({});
      const user = await UserFactory.create();
      const queue = await QueueFactory.create({
        allowQuestions: true,
        course: course,
      });
      await StudentCourseFactory.create({
        userId: user.id,
        courseId: queue.courseId,
      });
      await QuestionFactory.create({
        queueId: queue.id,
        creator: user,
        status: OpenQuestionStatus.Drafting,
      });

      const response = await postQuestion(user, queue);

      expect(response.status).toBe(400);
      expect(response.body.message).toBe(
        "You can't create more than one question at a time.",
      );
    });
    it('force a question when one is already open', async () => {
      const course = await CourseFactory.create({});
      const user = await UserFactory.create();
      const queue = await QueueFactory.create({
        allowQuestions: true,
        course: course,
      });
      await StudentCourseFactory.create({
        userId: user.id,
        courseId: queue.courseId,
      });
      await QuestionFactory.create({
        queueId: queue.id,
        creator: user,
        status: OpenQuestionStatus.Drafting,
      });

      await postQuestion(user, queue, true).expect(201);
    });
    it('lets student (who is TA in other class) create question', async () => {
      const user = await UserFactory.create();

      // Make user a TA in other class
      const queueOther = await QueueFactory.create({});
      await TACourseFactory.create({
        userId: user.id,
        courseId: queueOther.courseId,
      });

      // Make them student
      const queue = await QueueFactory.create({ allowQuestions: true });
      await StudentCourseFactory.create({
        userId: user.id,
        courseId: queue.courseId,
      });

      await postQuestion(user, queue).expect(201);
    });
    it('works when other queues and courses exist', async () => {
      const user = await UserFactory.create();

      await QueueFactory.create({});

      // Make them student
      const queue = await QueueFactory.create({ allowQuestions: true });
      await StudentCourseFactory.create({
        userId: user.id,
        courseId: queue.courseId,
      });

      await postQuestion(user, queue).expect(201);
    });
  });

  describe('PATCH /questions/:id', () => {
    it('as student creator, edit a question', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({ courseId: course.id });
      const user = await UserFactory.create();
      await StudentCourseFactory.create({ user, courseId: queue.courseId });
      const q = await QuestionFactory.create({
        text: 'Help pls',
        queue: queue,
        creator: user,
      });

      const response = await supertest({ userId: q.creatorId })
        .patch(`/questions/${q.id}`)
        .send({
          text: 'NEW TEXT',
        })
        .expect(200);
      expect(response.body).toMatchObject({
        id: q.id,
        text: 'NEW TEXT',
      });
      expect(await QuestionModel.findOne({ id: q.id })).toMatchObject({
        text: 'NEW TEXT',
      });
    });
    it('fails to update a non-existent question', async () => {
      await supertest({ userId: 99 })
        .patch(`/questions/999`)
        .send({
          text: 'NEW TEXT',
        })
        .expect(404);
    });
    it('PATCH taHelped as student is not allowed', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({ courseId: course.id });
      const student = await UserFactory.create();
      await StudentCourseFactory.create({
        user: student,
        courseId: queue.courseId,
      });

      const q = await QuestionFactory.create({
        text: 'Help pls',
        creator: student,
        queue: queue,
      });

      const ta = await UserFactory.create();
      await TACourseFactory.create({ course: q.queue.course, user: ta });

      await supertest({ userId: q.creatorId })
        .patch(`/questions/${q.id}`)
        .send({
          taHelped: {
            id: ta.id,
            name: ta.name,
          },
        })
        .expect(400);
    });
    it('PATCH status to helping as student not allowed', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({ courseId: course.id });
      const student = await UserFactory.create();
      await StudentCourseFactory.create({
        user: student,
        courseId: queue.courseId,
      });
      const q = await QuestionFactory.create({
        text: 'Help pls',
        creator: student,
        queue: queue,
      });

      await supertest({ userId: q.creatorId })
        .patch(`/questions/${q.id}`)
        .send({
          status: QuestionStatusKeys.Helping,
        })
        .expect(401);
    });
    it('PATCH status to helping as TA works', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({ courseId: course.id });
      const q = await QuestionFactory.create({
        text: 'Help pls',
        queue: queue,
      });
      const ta = await UserFactory.create();
      await TACourseFactory.create({ courseId: queue.courseId, user: ta });

      const res = await supertest({ userId: ta.id })
        .patch(`/questions/${q.id}`)
        .send({
          status: QuestionStatusKeys.Helping,
        })
        .expect(200);
      expect(res.body).toMatchObject({
        status: QuestionStatusKeys.Helping,
        taHelped: { id: ta.id, name: ta.name, photoURL: ta.photoURL },
      });
      expectUserNotified(q.creatorId);
    });
    it('PATCH status to Resolved as TA works', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({ courseId: course.id });
      const ta = await UserFactory.create();
      await TACourseFactory.create({ courseId: queue.courseId, user: ta });

      const q = await QuestionFactory.create({
        text: 'Help pls',
        status: QuestionStatusKeys.Helping,
        queue: queue,
        taHelped: ta,
      });

      const res = await supertest({ userId: ta.id })
        .patch(`/questions/${q.id}`)
        .send({
          status: QuestionStatusKeys.Resolved,
        })
        .expect(200);
      expect(res.body).toMatchObject({
        status: QuestionStatusKeys.Resolved,
      });
    });
    it('PATCH anything other than status as TA not allowed', async () => {
      const q = await QuestionFactory.create({ text: 'Help pls' });
      const ta = await UserFactory.create();
      await TACourseFactory.create({ course: q.queue.course, user: ta });

      await supertest({ userId: ta.id })
        .patch(`/questions/${q.id}`)
        .send({
          text: 'bonjour',
        })
        .expect(401);
    });
    it('PATCH invalid state transition not allowed', async () => {
      const q = await QuestionFactory.create({ text: 'Help pls' });
      const ta = await UserFactory.create();
      await TACourseFactory.create({ course: q.queue.course, user: ta });

      const res = await supertest({ userId: ta.id })
        .patch(`/questions/${q.id}`)
        .send({
          status: ClosedQuestionStatus.StudentCancelled
        })
        .expect(401);
      expect(res.body?.message).toContain("TA cannot change status from ")
    });
    it('PATCH question fails when you are not the question creator', async () => {
      const q = await QuestionFactory.create({ text: 'Help pls' });
      const student = await UserFactory.create();
      await StudentCourseFactory.create({
        course: q.queue.course,
        user: student,
      });

      await supertest({ userId: student.id })
        .patch(`/questions/${q.id}`)
        .send({
          text: 'bonjour',
        })
        .expect(401);
    });
    it('User not in course cannot patch a question', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({ courseId: course.id });
      const user = await UserFactory.create();
      const q = await QuestionFactory.create({
        text: 'Help pls',
        queue: queue,
        creator: user,
      });

      await supertest({ userId: q.creatorId })
        .patch(`/questions/${q.id}`)
        .send({
          text: 'NEW TEXT',
        })
        .expect(404); // Don't leak that course exists
    });
    it('Tries to help more than one student', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({ courseId: course.id });
      const q1 = await QuestionFactory.create({
        text: 'Help pls',
        queue: queue,
      });
      const q2 = await QuestionFactory.create({
        text: 'Help pls 2',
        queue: queue,
      });
      const ta = await UserFactory.create();
      await TACourseFactory.create({ courseId: queue.courseId, user: ta });

      await supertest({ userId: ta.id })
        .patch(`/questions/${q1.id}`)
        .send({
          status: QuestionStatusKeys.Helping,
        })
        .expect(200);
      await supertest({ userId: ta.id })
        .patch(`/questions/${q2.id}`)
        .send({
          status: QuestionStatusKeys.Helping,
        })
        .expect(400);
    });
  });
});
