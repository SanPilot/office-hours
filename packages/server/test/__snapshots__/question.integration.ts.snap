// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Integration GET /questions/:id gets a question with the given id 1`] = `
Object {
  "closedAt": null,
  "createdAt": "2020-03-01T05:00:00.000Z",
  "creator": Object {
    "email": "user1@neu.edu",
    "id": 1,
    "name": "John Doe the 1th",
    "photoURL": "https://pics/1",
  },
  "helpedAt": null,
  "id": 1,
  "isOnline": null,
  "location": null,
  "questionType": "Other",
  "status": "Queued",
  "taHelped": null,
  "text": "Help pls",
}
`;
